### PR TITLE
mgr/dashboard: Configure subsystems from dashboard 

### PR DESCRIPTION
**Pull request by afreen23**
Fixes https://tracker.ceph.com/issues/66659

- adds subsytems tab
- adds subsystem listing view
- adds create subsystem modal
- adds delete subsystem
- adds unit tests

[Screencast from 2024-06-25 06-03-50.webm](https://github.com/ceph/ceph/assets/25664409/5b37f84b-42a6-4994-8372-fb12f29f8af6)

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins retest this please

**Review by nizamial09 — APPROVED** [review]
Thank you @afreen23. code looks good to me and thank you for adding the unit tests.

Also thanks for using the routed modal for opening up the forms.. I also prefer the routed one over the normal which can be closed accidentally by clicking outside.

let some minor comments.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  ngOnInit() {
+    this.createForm();
+    this.action = this.actionLabels.CREATE;
+  }
+
+  createForm() {
+    this.subsystemForm = new CdFormGroup({
+      nqn: new UntypedFormControl('nqn.2001-07.com.ceph:' + Date.now(), {
```
[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:50]: I just copied from iscsi one, the format of NQN is to have "nqn":<period domain registered>:<domain> ...
Therefore I assume its time period when domain ceph.com perhaps registered.

User can still edit this field and replace it with its own NQN, this is default we have in UI for quick creation.

[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:50]: Here are references:
https://github.com/ceph/ceph/blob/96ef39027523af13cdfed07c34726abe224ac5d7/src/pybind/mgr/dashboard/frontend/src/app/ceph/block/iscsi-target-form/iscsi-target-form.component.ts#L182

https://nvmexpress.org/wp-content/uploads/NVMe_Over_Fabrics.pdf
From spec https://nvmexpress.org/wp-content/uploads/NVM-Express-Base-Specification-2.0c-2022.10.04-Ratified-1.pdf
```
 This format consists of:
• The string “nqn”;
• The string “.” (i.e., the ASCII period character);
• A date code, in “yyyy-mm” format. This date shall be during a time interval when the naming
authority owned the domain name used in this format. The date code uses the Gregorian
calendar. All digits and the dash shall be included;
• The string “.” (i.e., the ASCII period character);
• The reverse domain name of the naming authority that is creating the NQN; and
• A colon (:) prefixed string that the owner of the domain name assigns that does not exceed the
maximum length. The naming authority is responsible to ensure that the NQN is worldwide
unique
````

[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:50]: Moreover just for check did a `whois` `ceph.com` indeed registered on `2001-07`

```
Domain Name: ceph.com
Registry Domain ID: 75121254_DOMAIN_COM-VRSN
Registrar WHOIS Server: whois.comlaude.com
Registrar URL: https://www.comlaude.com
Updated Date: 2024-06-22T17:52:41Z
Creation Date: 2001-07-18T10:57:22Z  <<<<<
```

[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:50]: ah okay. thanks for the explanation here.

[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:50]: let's follow the same then

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        ]
+      }),
+      max_namespaces: new UntypedFormControl(256, {
+        validators: [CdValidators.number(false), Validators.max(256), Validators.min(1)]
+      })
+    });
+  }
+
+  onSubmit() {
+    const self = this;
```
[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:70]: `this` is already used for the form therefore `self`, any other suggestions ?

**Comment by afreen23** [comment]
Recent update summary:

- fixed unit tests

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:70]: need to use in task wrapper service, just check below in same OnSubmit() function:

```
 this.taskWrapperService
      .wrapTaskAroundCall({
        task: new FinishedTask(taskUrl, {
          nqn: nqn
        }),
        call: this.nvmeofService.createSubsystem(request)
      })
      .subscribe({
        error() {
          self.subsystemForm.setErrors({ cdSubmitButton: true });
        },
        complete: () => {
          this.router.navigate([this.pageURL, { outlets: { modal: null } }]);
        }
  ```

[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts]: :D changed now to "component"

**Comment by afreen23** [comment]
Recent update summary:

rename "self" to "component"

**Comment by afreen23** [comment]
jenkins test make check

**Review by aaSharma14 — APPROVED** [review]
Tested locally, Works fine. Thanks @afreen23

**Pull request merged**